### PR TITLE
Track timeseries with huge tagsets

### PR DESCRIPTION
**Pull request by djmitche**
This adds some telemetry and expvar exposure for counts of timeseries with huge, and almost-huge, tagsets.

### Motivation

This will help troubleshoot timeseries that are being over-labeled, which can sometimes cause operational issues.

### Describe how to test your changes

* Artificially inject some timeseries with lots and lots of tags
* See those counted in `expvar`
* See those counted in telemetry

I used
```
docker run --net=host --rm -ti --name agent -v $(pwd)/agent_telemetry:/etc/datadog-agent/conf.d/agent_telemetry:ro -v /home/dustin/tmp/integrations-internal/agent_telemetry/datadog_checks/agent_telemetry:/etc/datadog-agent/checks.d/agent_telemetry:ro -v /var/run/docker.sock:/var/run/docker.sock:ro -v /proc/:/host/proc/:ro -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro -e DD_API_KEY=*** -e DD_HOSTNAME=dustin-mitchell-tereshkova -e DD_TELEMETRY_ENABLED=true -l com.datadoghq.ad.check_names='["agent_telemetry","system_core"]' -l com.datadoghq.ad.init_configs="[{},{}]" -l com.datadoghq.ad.instances='[{"prometheus_url":"http://localhost:5000/telemetry","namespace":"datadog", "metrics": ["*"]},{"tags":["test:0", "test:1", "test:2", "test:3", "test:4", "test:5", "test:6", "test:7", "test:8", "test:9", "test:10", "test:11", "test:12", "test:13", "test:14", "test:15", "test:16", "test:17", "test:18", "test:19", "test:20", "test:21", "test:22", "test:23", "test:24", "test:25", "test:26", "test:27", "test:28", "test:29", "test:30", "test:31", "test:32", "test:33", "test:34", "test:35", "test:36", "test:37", "test:38", "test:39", "test:40", "test:41", "test:42", "test:43", "test:44", "test:45", "test:46", "test:47", "test:48", "test:49", "test:50", "test:51", "test:52", "test:53", "test:54", "test:55", "test:56", "test:57", "test:58", "test:59", "test:60", "test:61", "test:62", "test:63", "test:64", "test:65", "test:66", "test:67", "test:68", "test:69", "test:70", "test:71", "test:72", "test:73", "test:74", "test:75", "test:76", "test:77", "test:78", "test:79", "test:80", "test:81", "test:82", "test:83", "test:84", "test:85", "test:86", "test:87", "test:88", "test:89", "test:90", "test:91", "test:92", "test:93", "test:94", "test:95", "test:96", "test:97", "test:98", "test:99"]}]' -e DD_LOG_LEVEL=warn public.ecr.aws/datadog/agent:7.31.0-rc.4
```

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

### CODE REVIEW

File: pkg/aggregator/aggregator.go
Comment (hush-hush) [inline pkg/aggregator/aggregator.go:42]:
```suggestion
// hugeSeriesTagThreshold is the number of tags we consider "huge" and a sign
// that something is probably wrong with the series. If changing this value,
// also change the description in tlmHugeSeries.
```

File: pkg/aggregator/aggregator.go
Comment (hush-hush) [inline pkg/aggregator/aggregator.go:47]:
```suggestion
// almostHugeSeriesTagThreshold is a little less than HugeSeriesTagCount, serving as a
// threshold for series that are "almost" huge. If changing this value, also change
// the description in tlmAlmostHugeSeries.
```

File: pkg/aggregator/aggregator.go
Comment (hush-hush) [inline pkg/aggregator/aggregator.go:43]: Do we want to make those configurable ?

File: pkg/aggregator/aggregator.go
Comment (hush-hush) [inline pkg/aggregator/aggregator.go:55]:
```suggestion
// hugeSeriesCount is the total count of huge metric series. Access must be atomic.
var hugeSeriesCount uint64

// almostHugeSeriesCount is the total count of almost-huge metric series. Access must be atomic.
var almostHugeSeriesCount uint64

```

**Review by hush-hush — DISMISSED** [review]
Added a few nit-pick, looks good otherwise

File: pkg/aggregator/aggregator.go
Comment (djmitche) [reply pkg/aggregator/aggregator.go]: We might have to agree to disagree on this, or add a lint for it.  When I learned to type, full-stops were followed by two spaces.  I'll change it in this PR, but I don't expect I can manage to avoid doing this every time I write a sentence.  Also [the two spaces were useful in a monospaced context](https://www.quora.com/In-the-past-periods-full-stops-were-followed-by-two-spaces-Lately-its-become-common-to-see-just-one-space-In-the-modern-world-should-we-use-two-spaces-between-sentences-or-just-one-is-okay-Does-it-depend-on-the?share=1) and this is a monospace context :)

File: pkg/aggregator/aggregator.go
Comment (djmitche) [reply pkg/aggregator/aggregator.go]: @olivielpeau -- what do you think?

Comment (olivielpeau) [reply pkg/aggregator/aggregator.go]: No I don't think we should make these configurable, this is just for troubleshooting.

File: pkg/aggregator/aggregator.go
Comment (olivielpeau) [inline pkg/aggregator/aggregator.go:158]: I'd put the limit in the the metric name to be more descriptive, and make it pick up the constant value, i.e.:

```suggestion
	tlmHugeSeries = telemetry.NewCounter("aggregator", fmt.Sprintf("series_tags_above_%d", hugeTagsetThreshold),
		nil, fmt.Sprintf("Count of timeseries with over %d tags", hugeTagsetThreshold))
```

File: pkg/aggregator/aggregator.go
Comment (olivielpeau) [inline pkg/aggregator/aggregator.go:163]: Let's name this "sketch" instead of "distribution", to be consistent with all the other telemetry we expose for sketches/distributions.

Let's add the expvars as well.

**Comment by djmitche** [comment]
I refactored this to avoid the `almostHuge`/`huge` duality everywhere, and just use an array of thresholds.  It's not much shorter, but a bit cleaner!

File: pkg/aggregator/aggregator.go
Comment (remeh) [inline pkg/aggregator/aggregator.go:46]: Having a const representing "how many" thresholds we want to handle in this new logic may not be the best solution here IMO, especially since I'm not sure we'll change this number any time soon (and it would not be a big change anyway).
Defining your objects this way forces you to use this const everywhere for the slices and for the implementation 
 and it also results in duplicated ~15 lines of logic in `updateHugeSketchesTelemetry` and `updateHugeSeriesTelemetry`.

My point is that I think it is worth having a new type composed of all of these instead of all these different "independent" vars, it may be easier to read with this type also handling the logic.
These thresholds could be either provided as parameters while creating the instance (you would even be able to handle a dynamic amount of thresholds if you feel it's needed) or even hard-coded as const since they'll most likely be this way for a long time.

Not a strong opinion though.

File: pkg/aggregator/aggregator.go
Comment (remeh) [inline pkg/aggregator/aggregator.go:40]: It's way more than that, isn't it? Maybe something more generic like "tagsetSizeThresholdsCount is how many thresholds we want telemetry for."?

**Review by remeh — COMMENTED** [review]
Left a couple comments.

File: pkg/aggregator/aggregator.go
Comment (djmitche) [reply pkg/aggregator/aggregator.go:46]: The advantage was that I could use arrays for the globals, rather than slices.  But you're right, it's minor.  I'll refactor into a type.

File: pkg/aggregator/tagsest_telem.go
Comment (remeh) [inline pkg/aggregator/tagsest_telem.go:73]: You can maybe factorize that part which is the same in `updateHugeSeriesTelemetry` (only the counters/expvars are different)?

Comment (djmitche) [reply pkg/aggregator/tagsest_telem.go:73]: I wonder if we should consider a `tagset.Taggable` interface, for objects that are tagged, with `Tags() tagset.TagSet`?  That still wouldn't help too much here, since we'd have to iterate to turn `[]metrics.SketchSerie` into `[]tagset.Taggable`, but something to think about..

**Comment by djmitche** [comment]
(I'll take a review from either @remeh or @olivielpeau; not necessarily both)

**Review by remeh — COMMENTED** [review]
LGTM! Please make sure those failings tests aren't actual errors before merging.

**Review by remeh — APPROVED** [review]

**Comment by djmitche** [comment]
Oh, it's a simple golangci-lint issue...

**Comment by djmitche** [comment]
(I'll merge when green)

**Comment by djmitche** [comment]
TIL cmd-shift-enter will close a PR..

**Pull request merged**
